tests/test_data.py

```diff
@@ -2,198 +2,537 @@
+        u = TimeFunction(name='yu4D', grid=grid, space_order=0, time_order=1)
+        u.data[:] = 0.
+
+        # Test slicing w/ negative indices, combined to explicit indexing
+        u.data[1, 1:-1, 1:-1, 1:-1] = 6.
+        assert np.all(u.data[0] == 0.)
+        assert np.all(u.data[1, 1:-1, 1:-1, 1:-1] == 6.)
+        assert np.all(u.data[1, :, 0] == 0.)
+        assert np.all(u.data[1, :, -1] == 0.)
+        assert np.all(u.data[1, :, :, 0] == 0.)
+        assert np.all(u.data[1, :, :, -1] == 0.)
+
+    def test_halo_indexing(self):
+        """
+        Tests packing/unpacking data in :class:`Function` objects when some halo
+        region is present.
+        """
+        domain_shape = (16, 16, 16)
+        grid = Grid(shape=domain_shape)
+        u = Function(name='yu3D', grid=grid, space_order=2)
+
+        assert u.shape == u.data.shape == domain_shape
+        assert u.shape_with_halo == u.data_with_halo.shape == (20, 20, 20)
+
+        # Test simple insertion and extraction
+        u.data_with_halo[0, 0, 0] = 1.
+        u.data[0, 0, 0] = 2.
+        assert u.data_with_halo[0, 0, 0] == 1.
+        assert u.data[0, 0, 0] == 2.
+        assert u.data_with_halo[2, 2, 2] == 2.
+
+        # Test negative indices
+        u.data_with_halo[-1, -1, -1] = 3.
+        assert u.data[-1, -1, -1] == 0.
+        assert u.data_with_halo[-1, -1, -1] == 3.
+
+    def test_data_arithmetic(self):
+        """
+        Tests arithmetic operations between :class:`Data` objects and values.
+        """
+        grid = Grid(shape=(16, 16, 16))
+        u = Function(name='yu3D', grid=grid, space_order=0)
+        u.data[:] = 1
+
+        # Simple arithmetic
+        assert np.all(u.data == 1)
+        assert np.all(u.data + 2. == 3.)
+        assert np.all(u.data - 2. == -1.)
+        assert np.all(u.data * 2. == 2.)
+        assert np.all(u.data / 2. == 0.5)
+        assert np.all(u.data % 2 == 1.)
+
+        # Increments and partial increments
+        u.data[:] += 2.
+        assert np.all(u.data == 3.)
+        u.data[9, :, :] += 1.
+        assert all(np.all(u.data[i, :, :] == 3.) for i in range(9))
+        assert np.all(u.data[9, :, :] == 4.)
+
+        # Right operations __rOP__
+        u.data[:] = 1.
+        arr = np.ndarray(shape=(16, 16, 16), dtype=np.float32)
+        arr.fill(2.)
+        assert np.all(arr - u.data == 1.)
+
+    @skipif_yask  # YASK not throwing excpetions yet
+    def test_illegal_indexing(self):
+        """
+        Tests that indexing into illegal entries throws an exception.
+        """
+        nt = 5
+        grid = Grid(shape=(4, 4, 4))
+        u = Function(name='u', grid=grid)
+        v = TimeFunction(name='v', grid=grid, save=nt)
+
+        try:
+            u.data[5]
+            assert False
+        except IndexError:
+            pass
+        try:
+            v.data[nt]
+            assert False
+        except IndexError:
+            pass
+
+    def test_logic_indexing(self):
```

Comment:
Not sure about the name. Aren't logic indices `v_mod[v_mod > 0]` for example?

Reply:
it's the same name as before, I just put the pre-existing tests inside a class 

logic indexing is becase if your time buffer is of size 2, then you can access v[1], but also v[3], v[5], v[7], v[9], ...and they would all retrieve the same entry